### PR TITLE
Sort books in read menu

### LINKED ISSUE #33555: Read Menu Auto-Sort
**Is your feature request related to a problem? Please describe.**  
When I bunk down to read my library loot, I'm forever having to scroll down past books I've already read, or books that are just for fun, in order to get to the skill books. This is annoying.

**Describe the solution you'd like**  
The "Read" menu should be auto-sorted, like the Eat menu. First, list books that haven't been skimmed yet (your character doesn't know what's in them). Then sort the rest of the books: First, books that you can read to increase your skills; then other books, in order of most to least enjoyable; then books you haven't got the skill to understand yet.

### PR DESCRIPTION
**Pull request by OzoneH3**
#### Summary
```SUMMARY: Interface "Sort books in read menu"```

#### Purpose of change
Fixes #33555

### CODE REVIEW

File: src/game_inventory.cpp [hunk 1]
```diff
@@ -1004,6 +1004,35 @@ class read_inventory_preset: public pickup_inventory_preset
...
+                return ( !known_a && !known_b ) ? base_sort : !known_a;
+            }
+
+            const auto &book_a = get_book( lhs.any_item() );
+            const auto &book_b = get_book( rhs.any_item() );
+
+            if( !book_a.skill && !book_b.skill ) {
+                return ( book_a.fun == book_b.fun ) ? base_sort : book_a.fun > book_b.fun;
+            } else if( !book_a.skill || !book_a.skill ) {
+                return book_a.skill ? true : false;
```
Comment (BevapDin) [inline src/game_inventory.cpp]:
```suggestion
                return book_a.skill.has_value();
```

File: src/game_inventory.cpp [hunk 2]
```diff
@@ -1004,6 +1004,35 @@ class read_inventory_preset: public pickup_inventory_preset
...
+            }
+
+            const bool train_a = p.get_skill_level_object( book_a.skill ).level() < book_a.level;
```
Comment (BevapDin) [inline src/game_inventory.cpp]:
```suggestion
            const bool train_a = p.get_skill_level( book_a.skill ) < book_a.level;
```
Similar for the line below.

File: src/game_inventory.cpp (hunk 1 above)
Comment (OzoneH3) [reply src/game_inventory.cpp]: Hmm, doesn't have has_value() which seems to be a cata::optional thing but i get back a string_id<>.

Possible alternatives:
!book_a.skill.is_null()
!!book_a.skill

Comment (BevapDin) [reply src/game_inventory.cpp]: Stupid null ids. Use `!book_a.skill.is_null()` in that case.

Comment (OzoneH3) [reply src/game_inventory.cpp]: I settled for static_cast<bool>() which uses the skill_id operator just in case.

File: src/game_inventory.cpp [hunk 3]
```diff
@@ -1004,6 +1004,35 @@ class read_inventory_preset: public pickup_inventory_preset
...
+                return ( !known_a && !known_b ) ? base_sort : !known_a;
+            }
+
+            const auto &book_a = get_book( lhs.any_item() );
+            const auto &book_b = get_book( rhs.any_item() );
+
+            if( !book_a.skill && !book_b.skill ) {
+                return ( book_a.fun == book_b.fun ) ? base_sort : book_a.fun > book_b.fun;
+            } else if( !book_a.skill || !book_a.skill ) {
+                return static_cast<bool>(book_a.skill);
```
Comment (ZhilkinSerg) [inline src/game_inventory.cpp]:
```suggestion
                return static_cast<bool>( book_a.skill );
```

File: src/game_inventory.cpp [hunk 4]
```diff
@@ -1004,6 +1004,35 @@ class read_inventory_preset: public pickup_inventory_preset
...
+            if( !known_a || !known_b ) {
+                return ( !known_a && !known_b ) ? base_sort : !known_a;
+            }
+
+            const auto &book_a = get_book( lhs.any_item() );
+            const auto &book_b = get_book( rhs.any_item() );
+
+            if( !book_a.skill && !book_b.skill ) {
+                return ( book_a.fun == book_b.fun ) ? base_sort : book_a.fun > book_b.fun;
+            } else if( !book_a.skill || !book_a.skill ) {
```
Comment (ZhilkinSerg) [inline src/game_inventory.cpp]: Isn't there an error?

```suggestion
            } else if( !book_a.skill || !book_b.skill ) {
```

**Pull request merged**